File: 0046-permutations/0046-permutations.py
class Solution(object):
    def permute(self, nums):
          # PLAN OF ATTACK(POA)
        # 1) Understand the problem(DONE)
        #2) Understand a permutation(DONE)
         #3) Utilize recursion and setup permutations 
          #4)  remove the first num on list
          #5)Do a recursive call on perms with the modified nums
           #6) Use extend to add perms to perms result and append popped value to perms(Remember, it will be called recursively and execute again)
            # 7) return the perm_result outside the nested for loops.  
        
        #STEP ONE: define the result with the number of permutations
        perm_result = []
        
        # STEP TWO: define a base case for recursion
        if (len(nums) == 1):
            return [nums[:]]
        
        # STEP THREE: iterate through every value of nums via for loop
        
        for i in range(len(nums)):
            
            # remove the first value from nums
            n = nums.pop(0)
            
            # create a recursive call(note we have one less num due to it being popped)
            perms = self.permute(nums)
            
            # STEP FOUR: iterate through remaining perms and add it to the result
            for perm in perms:
                # adding popped value back onto permutation
                perm.append(n)
            # add the permutations back to perm results
            perm_result.extend(perms)
            
            nums.append(n)
            
            
        return perm_result
          
        
        
     
        

File: 0046-permutations/0046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
//     PLAN OF ATTACK
    
    
    // STEP ONE: Define the result array to store permutations
    let permResult = [];
    
    //STEP TWO: Base case for recursion: if there's only one element, return it
    if (nums.length === 1) {
        return [nums.slice()]; // Return a copy of nums as a single-element array
    }
    
    // STEP THREE: Iterate through each value of nums using a for loop
    for (let i = 0; i < nums.length; i++) {
        // Remove the first value from nums and store it in the variable n
        let n = nums.shift();
        
        // STEP FOUR: Recursive call to permute with the modified nums (one less element since it was shifted)
        let perms = permute(nums.slice());
        
        //STEP FIVE: Iterate through remaining permutations and add the shifted value back to each permutation
        for (let perm of perms) {
            perm.push(n); // Add the shifted value back to the permutation
        }
        
        // STEP SIX: Concatenate the updated permutations to permResult
        permResult.push(...perms);
        
        //STEP SEVEN:  Add the shifted value back to nums for the next iteration of the loop
        nums.push(n);
    }
    
    // Return the final permResult containing all permutations
    return permResult;
};

// Example usage
// const nums = [1, 2, 3];
// const result = permute(nums);
// console.log(result); // Output: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
